projects/plugins/jetpack/json-endpoints/class.wpcom-json-api-list-users-endpoint.php

```diff
@@ -164,87 +164,108 @@ public function callback( $path = '', $blog_id = 0 ) {
+						$search_term = str_replace( '*', '', $args['search'] );
+						return ( str_contains( $wp_viewer->user_login, $search_term ) || str_contains( $wp_viewer->user_email, $search_term ) || str_contains( $wp_viewer->display_name, $search_term ) );
+					}
+				);
+			}
+
+			$viewer_ids     = array();
+			$unique_viewers = array();
+			// Create a lookup array of viewer IDs.
+			foreach ( $viewers as $viewer ) {
+				$viewer_ids[ $viewer->ID ] = true;
+				$viewer->roles[]           = 'viewer';
+			}
+
+			// Add viewer role to users who are also viewers.
+			foreach ( $users as $user ) {
+				if ( isset( $viewer_ids[ $user->ID ] ) && ! in_array( 'viewer', $user->roles, true ) ) {
+					$user->roles[] = 'viewer';
+					// Mark this user so we don't add them again.
+					$viewer_ids[ $user->ID ] = false;
+				}
+			}
+
+			// Add viewer role to viewers and remove duplicates.
+			foreach ( $viewers as $viewer ) {
+				if ( isset( $viewer_ids[ $viewer->ID ] ) && true === $viewer_ids[ $viewer->ID ] ) {
+					$viewer->roles[]  = 'viewer';
+					$unique_viewers[] = $viewer;
+				}
+			}
+
+			// Reassign the viewers array to the unique viewers array.
+			$viewers = $unique_viewers;
 		}
 
-		$return = array();
 		foreach ( array_keys( $this->response_format ) as $key ) {
 			switch ( $key ) {
 				case 'found':
 					$user_count = (int) $user_query->get_total();
-
-					$viewer_count = 0;
 					if ( $include_viewers ) {
-						if ( empty( $args['search'] ) ) {
-							$viewer_count = (int) get_count_private_blog_users( $blog_id );
-						} else {
-							$viewer_count = count( $viewers );
-						}
+						$viewer_count     = count( $viewers );
+						$response[ $key ] = $user_count + $viewer_count;
+					} else {
+						$response[ $key ] = $user_count;
 					}
-
-					$return[ $key ] = $user_count + $viewer_count;
 					break;
 				case 'users':
-					$users        = array();
-					$is_multisite = is_multisite();
-					foreach ( $user_query->get_results() as $u ) {
-						$the_user = $this->get_author( $u, true );
-						if ( $the_user && ! is_wp_error( $the_user ) ) {
-							$userdata        = get_userdata( $u );
-							$the_user->roles = ! is_wp_error( $userdata ) ? array_values( $userdata->roles ) : array();
-							if ( $is_multisite ) {
-								$the_user->is_super_admin = user_can( $the_user->ID, 'manage_network' );
-							}
-							$users[] = $the_user;
-						}
-					}
-
-					$combined_users = array_merge( $users, $viewers );
-
-					// When viewers are included, we ignore the order & orderby parameters.
 					if ( $include_viewers ) {
-						usort(
-							$combined_users,
-							function ( $a, $b ) {
-								return strcmp( strtolower( $a->name ), strtolower( $b->name ) );
-							}
-						);
+						$response[ $key ] = array_merge( $users, $viewers );
+					} else {
+						$response[ $key ] = $users;
 					}
-
-					$return[ $key ] = $combined_users;
 					break;
 			}
 		}
 
-		return $return;
+		return $response;
```

Comment:
Something about these changes also screws with the `/people/viewers/thingsireadonthetoilet.wordpress.com/$user_id` Calypso page. Check.

Reply:
It's this check: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/people/team-members/index.tsx#L114

It assumes that the roles array is empty for viewers.

Reply:
Also watch out for double badges here:

https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/people/people-profile/index.jsx#L320-L321

<img width="772" alt="Screenshot 2025-02-27 at 5 46 05 pm" src="https://github.com/user-attachments/assets/43a609c7-fe39-41af-b2e7-179b4d4670c1" />